bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java

```diff
@@ -0,0 +1,793 @@
+
+    private static final String ZONE = "ZONE";
+    private static final String SOURCE = "SOURCE";
+    private static final String CHANNEL_DELIMIT = "#";
+    private static final String UNDEF = "UNDEF";
+    private static final String GC_STR = "NV-IG8";
+
+    private static final int MAX_ZONES = 20;
+    private static final int MAX_SRC = 6;
+    private static final int MIN_VOLUME = 0;
+    private static final int MAX_VOLUME = 79;
+    private static final int MIN_EQ = -18;
+    private static final int MAX_EQ = 18;
+
+    private static final Pattern ZONE_PATTERN = Pattern
+            .compile("^ON,SRC(\\d{1}),(MUTE|VOL\\d{1,2}),DND([0-1]),LOCK([0-1])$");
+    private static final Pattern DISP_PATTERN = Pattern.compile("^DISPLINE(\\d{1}),\"(.*)\"$");
+    private static final Pattern DISP_INFO_PATTERN = Pattern
+            .compile("^DISPINFO,DUR(\\d{1,6}),POS(\\d{1,6}),STATUS(\\d{1,2})$");
+    private static final Pattern ZONE_CFG_PATTERN = Pattern.compile("^BASS(.*),TREB(.*),BAL(.*),LOUDCMP([0-1])$");
+
+    private final Logger logger = LoggerFactory.getLogger(NuvoHandler.class);
+    private final NuvoStateDescriptionOptionProvider stateDescriptionProvider;
+    private final SerialPortManager serialPortManager;
+
+    private @Nullable ScheduledFuture<?> reconnectJob;
+    private @Nullable ScheduledFuture<?> pollingJob;
+    private @Nullable ScheduledFuture<?> clockSyncJob;
+
+    private NuvoConnector connector = new NuvoDefaultConnector();
+    private long lastEventReceived = System.currentTimeMillis();
+    private int numZones = 1;
+    private String versionString = BLANK;
+    private boolean isGConcerto = false;
+    private Object sequenceLock = new Object();
+
+    Set<Integer> activeZones = new HashSet<>(1);
+
+    // A state option list for the source labels
+    List<StateOption> sourceLabels = new ArrayList<>();
+
+    /**
+     * Constructor
+     */
+    public NuvoHandler(Thing thing, NuvoStateDescriptionOptionProvider stateDescriptionProvider,
+            SerialPortManager serialPortManager) {
+        super(thing);
+        this.stateDescriptionProvider = stateDescriptionProvider;
+        this.serialPortManager = serialPortManager;
+    }
+
+    @Override
+    public void initialize() {
+        NuvoThingConfiguration config = getConfigAs(NuvoThingConfiguration.class);
+        String serialPort = config.serialPort;
+        String host = config.host;
+        Integer port = config.port;
+        Integer numZones = config.numZones;
+
+        // Check configuration settings
+        String configError = null;
+        if ((serialPort == null || serialPort.isEmpty()) && (host == null || host.isEmpty())) {
+            configError = "undefined serialPort and host configuration settings; please set one of them";
+        } else if (serialPort != null && (host == null || host.isEmpty())) {
+            if (serialPort.toLowerCase().startsWith("rfc2217")) {
```

Comment:
RFC2217 through the openhab serial layer never worked in bindings I worked on. 
I am not sure of the reason but my hypothesis was it doesn't work if you need to read and write in parallel. 
But you are welcome to try with any new binding. Maybe I was just not enough clever to make it work.

Reply:
This is probably a problem in case you have a reader thread that always read data.

Reply:
That makes sense. But I think I will just stick with what works.

Reply:
@lolodomo Maybe it was related to the [endless sleep 'fix' (that could be and was removed during the review process)](https://github.com/openhab/openhab-addons/pull/5805/commits/7d559b6d1d911710861b9e303288919995341dfa)? Did you ever test after that change?

@mlobstein Did you test if it worked at all by using the rfc2217 version?

Reply:
@Hillbrand: Honestly I don't remember. I even don't remember if I tested that with this binding in particular. But as I am using serial over IP, I could try again one day.

Reply:
> 
> 
> @lolodomo Maybe it was related to the [endless sleep 'fix' (that could be and was removed during the review process)](https://github.com/openhab/openhab-addons/pull/5805/commits/7d559b6d1d911710861b9e303288919995341dfa)? Did you ever test after that change?
> 
> @mlobstein Did you test if it worked at all by using the rfc2217 version?
I did not test it with rfc2217

Reply:
If it's not tested or not known if the rfc2217 configuration way doesn't work I would be in favor of removing the specific ip implementation here and allowing the rfcc2217 configuratiin option for serial. Its always possible to add features, but removing them is hard. This also reduces the amount of copy-pasting of specific code.

Reply:
A couple questions...

1. I tried rfc2217 and it seemed to work connecting to a ser2net endpoint after a few changes to the serial connector. I had to conditionally bypass SerialPort.enableReceiveThreshold(), SerialPort.enableReceiveTimeout() & SerialPort.setFlowControlMode() as the calling those with rfc2217 throws an exception. I also had to catch an IllegalStateException that was being thrown if the remote connection was refused or otherwise failed to connect. That being said, are there any bindings that currently support rfc2217 explicitly (as a reference implementation)? I did a quick scan of the 2.5.x bindings and the only references to rfc2217 were in 6 other bindings that blocked its usage during initialization.

2. If rfc2217 is to be used, does the user need to make the choice to specify the rfc2217 url string in place of the serial port or can this be handled behind the scenes? There seems to be some controversy with setting limitToOptions=false for the serial port in the configuration. https://github.com/openhab/openhab-core/issues/1029 Once this is set to false, the user can type the url string for the rfc2217 connection (rfc2217://<ip>:<port>). I feel this is cumbersome and it would be much easier for the user to give them a place to specify IP and port # for the remote serial port and use those to construct the rfc2217 url in the absence of a physical serial port being selected from the drop down.
@kaikreuzer

Reply:
Thank you for extensive testing this. At least we know it works (besides the exception throwing. In know the dsmr binding does also works once the exceptions would be wrapped.). The openhab-core issue 1029 you mentioned was about the possibility to be able to add a network connection. It doesn't even address the usability of it, which is too bad. I also looked at the other bindings. It looks to me this has grown historically. In the past the gnu serial library was used directly, so no rfc2217 support was availble. So that earlier binding implemented their own version of it. Later versions seem to be made by the same person and he copied that implementation. The other bindings seem to add a bridge for ip configuration. I also noted we already approved your oppo binding that does the same.

So I'll guess it should be ok to leave it as is here. What I'm planning, is to see if I can get those quircks fixed in core so it behaves the same as other direct serial connections and than see if this makes it possible to remove the custom implementations for ip connections. For example here you could construct the rfc2217 connection fromt host-port input and then use the rfc2217 of core. The configurations can than still remain (with addition that rfc2217 configuration would be allowed on serialPort configuration). But that would not be available before openHAB 3.0.